addon/-private/system/relationships/relationships-payloads.js

```diff
@@ -0,0 +1,417 @@
+import Ember from 'ember';
+let get = Ember.get;
+
+/**
+  Manages relationship payloads for a given store.  Acts as a single source of
+  truth (of payloads) for both sides of a relationship so they can agree on the
+  most up-to-date payload received without needing too much eager processing
+  when those payloads are pushed into the store.
+
+  @example
+
+    let relationshipsPayloads = new RelationshipsPayloads(store);
+
+    const User = DS.Model.extend({
+      hobbies: DS.hasMany('hobby')
+    });
+
+    const Hobby = DS.Model.extend({
+      user: DS.belongsTo('user')
+    });
+
+    let userPayload = {
+      data: {
+        id: 1,
+        type: 'user',
+        relationships: {
+          hobbies: {
+            data: [{
+              id: 2,
+              type: 'hobby'
+            }]
+          }
+        }
+      },
+    };
+    relationshipsPayloads.push('user', 1, userPayload.data.relationships);
+
+    relationshipsPayloads.get('hobby', 2, 'user') === {
+      {
+        data: {
+          id: 1,
+          type: 'user'
+        }
+      }
+    }
+
+  @private
+  @class RelationshipsPayloads
+*/
+export default class RelationshipsPayloads {
+  constructor(store) {
+    this._store = store;
+    // cache of `RelationshipPayload`s
+    this._cache = Object.create(null);
+  }
+
+  /**
+    Find the payload for the given relationship of the given model.
+
+    Returns the payload for the given relationship, whether raw or computed from
+    the payload of the inverse relationship.
+
+    @example
+
+      relationshipsPayloads.get('hobby', 2, 'user') === {
+        {
+          data: {
+            id: 1,
+            type: 'user'
+          }
+        }
+      }
+
+    @method
+  */
+  get(modelName, id, relationshipName) {
+    let modelClass = this._store._modelFor(modelName);
+    let relationshipsByName = get(modelClass, 'relationshipsByName');
+    let relationshipPayloads = this._getRelationshipPayloads(modelName, relationshipName, modelClass, relationshipsByName, false);
+    return relationshipPayloads && relationshipPayloads.get(modelName, id, relationshipName);
+  }
+
+  /**
+    Push a model's relationships payload into this cache.
+
+    @example
+
+      let userPayload = {
+        data: {
+          id: 1,
+          type: 'user',
+          relationships: {
+            hobbies: {
+              data: [{
+                id: 2,
+                type: 'hobby'
+              }]
+            }
+          }
+        },
+      };
+      relationshipsPayloads.push('user', 1, userPayload.data.relationships);
+
+    @method
+  */
+  push(modelName, id, relationshipsData) {
+    if (!relationshipsData) { return; }
+
+    let modelClass = this._store._modelFor(modelName);
+    let relationshipsByName = get(modelClass, 'relationshipsByName');
+    for (let key in relationshipsData) {
+      let relationshipPayloads = this._getRelationshipPayloads(modelName, key, modelClass, relationshipsByName, true);
```

Comment:
unfortunately we'd need to also pass the inverse relationship, which we don't always need to compute